File: apps/api/src/personal_enigma/api/goose_pixels.py
"""C35 — Goose pixel licence: work presence, C34 expressiveness.

Local SURFACE projection. Never attach to the remote working set.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from personal_enigma.api.relational_bootstrap import RelationalBootstrapBlock

GooseMotion = Literal["absent", "idle", "walk", "return"]
GooseExpressiveness = Literal["restrained", "playful"]
AgentWorkPhase = Literal["in_flight", "waiting", "complete"]
VisibilityLayer = Literal["surface", "inspectable", "forensic"]

_PHASE_MOTION: dict[AgentWorkPhase, GooseMotion] = {
    "in_flight": "walk",
    "waiting": "idle",
    "complete": "return",
}


@dataclass(frozen=True, slots=True)
class AgentWorkSnapshot:
    """Real work already knowable on SURFACE — not a Goose dossier."""

    exists: bool
    phase: AgentWorkPhase | None = None
    semantic_token: str = ""
    inspect_target: str | None = None
    inspect_labels: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class GoosePixelLicence:
    motion: GooseMotion
    expressiveness: GooseExpressiveness
    layer: Literal["surface"] = "surface"
    grants_authority: bool = False
    is_evidence: bool = False
    inspect_target: str | None = None
    inspect_labels: tuple[str, ...] = ()
    work_semantic_token: str = ""

    def as_wire(self) -> dict[str, Any]:
        return {
            "kind": "goose_pixel_licence",
            "layer": self.layer,
            "motion": self.motion,
            "expressiveness": self.expressiveness,
            "grants_authority": False,
            "is_evidence": False,
            "inspect_target": self.inspect_target,
            "inspect_labels": list(self.inspect_labels),
            "work_semantic_token": self.work_semantic_token,
        }


def motion_from_work(work: AgentWorkSnapshot | None) -> GooseMotion:
    if work is None or not work.exists or work.phase is None:
        return "absent"
    return _PHASE_MOTION[work.phase]


def expressiveness_from_bootstrap(
    block: RelationalBootstrapBlock | None,
) -> GooseExpressiveness:
    if block is not None and block.culture_palette_available:
        return "playful"
    return "restrained"


def license_goose_pixels(
    *,
    work: AgentWorkSnapshot | None,
    bootstrap: RelationalBootstrapBlock | None,
) -> GoosePixelLicence:
    motion = motion_from_work(work)
    if motion == "absent":
        return GoosePixelLicence(motion="absent", expressiveness="restrained")
    assert work is not None
    return GoosePixelLicence(
        motion=motion,
        expressiveness=expressiveness_from_bootstrap(bootstrap),
        inspect_target=work.inspect_target,
        inspect_labels=work.inspect_labels,
        work_semantic_token=work.semantic_token,
    )


def pixels_allowed_on(layer: VisibilityLayer, licence: GoosePixelLicence) -> bool:
    if layer != "surface":
        return False
    return licence.motion != "absent"


__all__ = [
    "AgentWorkPhase",
    "AgentWorkSnapshot",
    "GoosePixelLicence",
    "expressiveness_from_bootstrap",
    "license_goose_pixels",
    "motion_from_work",
    "pixels_allowed_on",
]

File: apps/web/src/enigma/goosePixels.ts
import type { ConversationItem } from "./types";
import { threadActivityFromTrace, type EnigmaActivityEvent } from "./activity";

export type GooseMotion = "absent" | "idle" | "walk" | "return";
export type GooseExpressiveness = "restrained" | "playful";
export type AgentWorkPhase = "in_flight" | "waiting" | "complete";
export type VisibilityLayer = "surface" | "inspectable" | "forensic";

export type AgentWorkSnapshot = {
  exists: boolean;
  phase: AgentWorkPhase | null;
  semanticToken: string;
  inspectTarget: string | null;
  inspectLabels: string[];
};

export type GoosePixelLicence = {
  motion: GooseMotion;
  expressiveness: GooseExpressiveness;
  layer: "surface";
  grantsAuthority: false;
  isEvidence: false;
  inspectTarget: string | null;
  inspectLabels: string[];
  workSemanticToken: string;
};

const PHASE_MOTION: Record<AgentWorkPhase, GooseMotion> = {
  in_flight: "walk",
  waiting: "idle",
  complete: "return",
};

export function emptyWorkSnapshot(): AgentWorkSnapshot {
  return {
    exists: false,
    phase: null,
    semanticToken: "",
    inspectTarget: null,
    inspectLabels: [],
  };
}

export function motionFromWork(work: AgentWorkSnapshot | null): GooseMotion {
  if (!work?.exists || work.phase === null) {
    return "absent";
  }
  return PHASE_MOTION[work.phase];
}

export function expressivenessFromRemoteContext(remote: unknown): GooseExpressiveness {
  if (!remote || typeof remote !== "object") {
    return "restrained";
  }
  const bootstrap = (remote as { relational_bootstrap?: unknown }).relational_bootstrap;
  if (!bootstrap || typeof bootstrap !== "object") {
    return "restrained";
  }
  const continuation = (bootstrap as { continuation?: unknown }).continuation;
  if (!continuation || typeof continuation !== "object") {
    return "restrained";
  }
  return (continuation as { culture_palette_available?: unknown }).culture_palette_available === true
    ? "playful"
    : "restrained";
}

export function licenseGoosePixels(
  work: AgentWorkSnapshot | null,
  expressiveness: GooseExpressiveness,
): GoosePixelLicence {
  const motion = motionFromWork(work);
  if (motion === "absent" || !work) {
    return {
      motion: "absent",
      expressiveness: "restrained",
      layer: "surface",
      grantsAuthority: false,
      isEvidence: false,
      inspectTarget: null,
      inspectLabels: [],
      workSemanticToken: "",
    };
  }
  return {
    motion,
    expressiveness,
    layer: "surface",
    grantsAuthority: false,
    isEvidence: false,
    inspectTarget: work.inspectTarget,
    inspectLabels: work.inspectLabels,
    workSemanticToken: work.semanticToken,
  };
}

export function pixelsAllowedOn(layer: VisibilityLayer, licence: GoosePixelLicence): boolean {
  return layer === "surface" && licence.motion !== "absent";
}

export function latestThreadActivities(items: ConversationItem[]): EnigmaActivityEvent[] {
  for (let index = items.length - 1; index >= 0; index -= 1) {
    const item = items[index];
    const trace = item?.llm_trace;
    if (trace) {
      return threadActivityFromTrace(trace, { at: item.at });
    }
  }
  return [];
}

export function latestRemoteContext(items: ConversationItem[]): unknown {
  for (let index = items.length - 1; index >= 0; index -= 1) {
    const remote = items[index]?.llm_trace?.remote_context_sent;
    if (remote !== undefined && remote !== null) {
      return remote;
    }
  }
  return null;
}

export function workSnapshotFromConversation(input: {
  items: ConversationItem[];
  busy: boolean;
  loading: boolean;
}): AgentWorkSnapshot {
  const activities = latestThreadActivities(input.items);
  const labels = activities.map((event) => event.label);
  const inspectTarget = activities.find((event) => event.subject_id)?.subject_id ?? null;
  const pendingAssist = input.items.some((item) => item.kind === "assist_proposal")
    && !input.items.some((item) => item.kind === "assist_result");

  if (input.busy || input.loading) {
    return {
      exists: true,
      phase: "in_flight",
      semanticToken: activities[0]?.id ?? "in-flight",
      inspectTarget,
      inspectLabels: labels,
    };
  }

  if (pendingAssist) {
    const proposal = [...input.items]
      .reverse()
      .find((item): item is Extract<ConversationItem, { kind: "assist_proposal" }> => {
        return item.kind === "assist_proposal";
      });
    return {
      exists: true,
      phase: "waiting",
      semanticToken: proposal?.proposal.id ?? "assist-waiting",
      inspectTarget,
      inspectLabels: proposal ? [proposal.proposal.title] : ["Prepared an action"],
    };
  }

  if (activities.length > 0) {
    return {
      exists: true,
      phase: "complete",
      semanticToken: activities.map((event) => event.id).join("|"),
      inspectTarget,
      inspectLabels: labels,
    };
  }

  return emptyWorkSnapshot();
}

export function licenceFromConversation(input: {
  items: ConversationItem[];
  busy: boolean;
  loading: boolean;
}): GoosePixelLicence {
  return licenseGoosePixels(
    workSnapshotFromConversation(input),
    expressivenessFromRemoteContext(latestRemoteContext(input.items)),
  );
}
